receiver/snowflakereceiver/client.go

```diff
@@ -13,3 +13,426 @@
+	sessionMetricsQuery          = "select USER_NAME, count(distinct(SESSION_ID)) from Sessions where created_on >= DATEADD(hour, -24, current_timestamp()) group by 1;"
+	snowpipeMetricsQuery         = "select pipe_name, sum(credits_used), sum(bytes_inserted), sum(files_inserted) from pipe_usage_history where start_time >= DATEADD(hour, -24, current_timestamp()) group by 1;"
+	storageMetricsQuery          = "select STORAGE_BYTES, STAGE_BYTES, FAILSAFE_BYTES from STORAGE_USAGE ORDER BY USAGE_DATE DESC LIMIT 1;"
+)
+
+// snowflake client is comprised of a sql.DB (the proper 'client' in question),
+// a connection string (dsn), a collection of queries (built from which metrics are enabled),
+// and a logger
+type defaultSnowflakeClient struct {
+	client *sql.DB
+	dsn    *string
+	logger *zap.Logger
+}
+
+var _ snowflakeClient = (*defaultSnowflakeClient)(nil)
+
+type snowflakeClient interface {
+	GetBillingMetrics(ctx context.Context) (*[]billingMetric, error)
+	GetWarehouseBillingMetrics(ctx context.Context) (*[]whBillingMetric, error)
+	GetLoginMetrics(ctx context.Context) (*[]loginMetric, error)
+	GetHighLevelQueryMetrics(ctx context.Context) (*[]hlQueryMetric, error)
+	GetDbMetrics(ctx context.Context) (*[]dbMetric, error)
+	GetSessionMetrics(ctx context.Context) (*[]sessionMetric, error)
+	GetSnowpipeMetrics(ctx context.Context) (*[]snowpipeMetric, error)
+	GetStorageMetrics(ctx context.Context) (*[]storageMetric, error)
+}
+
+// build snowflake db connection string
+func buildDSN(cfg Config) string {
+	conf := &sf.Config{
+		Account:   cfg.Account,
+		User:      cfg.Username,
+		Password:  cfg.Password,
+		Database:  cfg.Database,
+		Schema:    cfg.Schema,
+		Role:      cfg.Role,
+		Warehouse: cfg.Warehouse,
+	}
+
+	dsn, err := sf.DSN(conf)
+	if err != nil {
+		print("%v", err)
+	}
+
+	return dsn
+}
+
+func newDefaultClient(settings component.TelemetrySettings, c Config) (*defaultSnowflakeClient, error) {
+	dsn := buildDSN(c)
+	db, err := sql.Open("snowflake", dsn)
+	if err != nil {
+		return nil, err
+	}
+
+	return &defaultSnowflakeClient{
+		client: db,
+		dsn:    &dsn,
+		logger: settings.Logger,
+	}, nil
+}
+
+// queries database and returns resulting rows
+func (c defaultSnowflakeClient) readDB(ctx context.Context, q string) (*sql.Rows, error) {
+	rows, err := c.client.QueryContext(ctx, q)
+	if err != nil {
+		error := fmt.Sprintf("Query failed with %v", err)
+		c.logger.Error(error)
+		return nil, err
+	}
+
+	return rows, nil
+}
+
+// these wrap readDB and return the associated data type which the scraper will
+// use to generate and emit metrics. Which of these are called will be based on which metrics
+// are enabled in the Config (default is all of them)
+func (c defaultSnowflakeClient) GetBillingMetrics(ctx context.Context) (*[]billingMetric, error) {
```

Comment:
nit: `Get` prefix is considered redundant in golang world, so maybe `billingMetrics` or `fetchBillingMetrics`.

Reply:
Thanks for the note on this, I was unaware!